File: go/internal/gameloop/geometry.go
package gameloop

type Rect struct {
	X      float64
	Y      float64
	Width  float64
	Height float64
}

func (r *Rect) SetPosition(x, y float64) {
	r.X = x
	r.Y = y
}

func (r1 *Rect) Collides(r2 *Rect) bool {
	if r1.X > r2.X+r2.Width || r2.X > r1.X+r1.Width {
		return false
	}
	if r1.Y > r2.Y+r2.Height || r2.Y > r1.Y+r1.Height {
		return false
	}
	return true
}

type Vector2D = [2]float64

File: ts/src/geometry.ts
export type Vector2D = [number, number];

export class Rect {
    x: number;
    y: number;
    width: number;
    height: number;

    constructor(x: number, y: number, width: number, height: number) {
        this.x = x;
        this.y = y;
        this.width = width;
        this.height = height;
    }

    setPosition(x: number, y: number): void {
        this.x = x;
        this.y = y;
    }

    collides(other: Rect): boolean {
        if (this.x > other.x + other.width || other.x > this.x + this.width) {
            return false;
        }
        if (this.y > other.y + other.height || other.y > this.y + this.height) {
            return false;
        }
        return true;
    }
}
